lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp

```diff
@@ -619,6 +856,131 @@ removeLocalNonNestedAccess(const AccessConflictAnalysis::Result &result,
   return changed;
 }
 
+// TODO: support multi-end access cases
+static EndAccessInst *getSingleEndAccess(BeginAccessInst *inst) {
+  EndAccessInst *end = nullptr;
+  for (auto *currEnd : inst->getEndAccesses()) {
+    if (end == nullptr)
+      end = currEnd;
+    else
+      return nullptr;
+  }
+  return end;
+}
+
+static void mergeEndAccesses(BeginAccessInst *parentIns,
+                             BeginAccessInst *childIns) {
+  auto *endP = getSingleEndAccess(parentIns);
+  if (!endP)
+    llvm_unreachable("not supported");
+  auto *endC = getSingleEndAccess(childIns);
+  if (!endC)
+    llvm_unreachable("not supported");
+
+  endC->setOperand(parentIns);
+  endP->eraseFromParent();
+}
+
+static bool canMergeEnd(BeginAccessInst *parentIns, BeginAccessInst *childIns) {
+  auto *endP = getSingleEndAccess(parentIns);
+  if (!endP)
+    return false;
+
+  auto *endC = getSingleEndAccess(childIns);
+  if (!endC)
+    return false;
+
+  return true;
+}
+
+// TODO: support other merge patterns
+static bool canMergeBegin(PostDominanceInfo *postDomTree,
+                          BeginAccessInst *parentIns,
+                          BeginAccessInst *childIns) {
+  if (!postDomTree->properlyDominates(childIns, parentIns)) {
+    return false;
+  }
+  return true;
+}
+
+static bool canMerge(PostDominanceInfo *postDomTree, BeginAccessInst *parentIns,
+                     BeginAccessInst *childIns) {
+  if (!canMergeBegin(postDomTree, parentIns, childIns))
+    return false;
+
+  return canMergeEnd(parentIns, childIns);
+}
+
+/// Perform access merging.
+static bool
+mergeAccesses(PostDominanceInfo *postDomTree,
+              const AccessConflictAndMergeAnalysis::MergeableMap &mergeMap) {
+  bool changed = false;
+
+  // make a temporary reverse copy to work on:
+  // It is in reverse order just to make it easier to debug / follow
+  AccessConflictAndMergeAnalysis::MergeableMap workMap;
+  workMap.append(mergeMap.rbegin(), mergeMap.rend());
+
+  // Assume we have two pairs in map (1,2) , (2,3)
```

Comment:
Also here, the result is not a map.
Actually the comment is a bit cryptic. How about:
"Assume the result contains two access pairs to be merged:
  (begin_access %1, begin_access %2) // = merge end_access %1 with begin_access %2
  (begin_access %2, begin_access %3) // = merge end_access %2 with begin_access %3 
After merging the first pair, begin_access %2 is removed, so the second pair in the result list points to a to-be-deleted begin_access instruction.  We store (begin_access %2 -> begin_access %1) to re-map a merged begin_access to it's replaced instruction."